docs/design/2019-04-11-indexmerge.md

```diff
@@ -0,0 +1,284 @@
+		results = append(results, indexPath)
+	}
+	return results
+}
+
+// This function will get a best indexPath for a con from some alternative paths.
+// now we just take the index which has more columns.
+// for exmple: 
+// (1)
+// index1(a,b,c) index2(a,b) index3(a)
+// condition: a = 1 will choose index1; a = 1 and b = 2 will also choose index1
+// (2)
+// index1(a) index2(b)
+// condition: a = 1 and b = 1
+// random choose???
+GetIndexMergePartialPath(IndexInfos, indexAccessPaths) {
+}
+
+// (1)maybe we will merge some indexPaths
+//    for example: index1(a) index2(b)
+//    condition : a < 1 or a > 2 or b < 1 or b > 10 
+//    imPaths will be [a<1,a>2,b<1,b>10] and we can merge it and get [a<1 or a >2 , b < 1 or b > 10] 
+// (2)IndexMergePath.tableFilters: 
+//    <1> remove cond from PushdownConditions and the remain will be added to tableFitler.
+//    <2> after merge operation, if any indexPath's tableFilter is not nil, we should add it into tableFilters
+	
+CreateIndexMergeUnionPath(imPaths,PushdownConditions,cond,IndexInfos) {
+}
+
+```
+
+
+```
+GetIndexMergeIntersectionPaths(pushDownConditions, usedConditionsInOr, indexInfos) {
+	var partialPaths
+
+	if len(pushDownConditions) - len(usedConditionsInOr) < 2 {
+	return nil
+}
+tableFilters := append(tableFilters, usedConditionsInOr...)
+newConsiderConditions := remove(pushDownConditions, usedConditionsInOr)
+for cond in newConsiderConditions {
+	indexPaths = buildAccessPath([]{cond}, indexInfos)
+	if indexPaths == nil {
+		tableFiltes = append(tableFilters,cond)
+		continue
+}
+indexPath := GetIndexMergePartialPath(indexPaths,indexInfos)
+partialPaths = append(partialPaths, indexPath)
+}
+if len(partialPaths) < 2 {
+	return nil
+}
+return CreateIndexMergeIntersectionPath(partialPaths, tableFilters)
+}
+
+// Now, we just use all path in partialPaths to generate a IndexMergeIntersection.
+// We also need to merge possible paths. 
+// For example: 
+//   index:  ix1(a)
+//   condition： a > 1 and a < 10
+//   we will get two partial paths and they all use index ix1. 
+// IndexMergePath.tableFilters: 
+//   <1> tableFilters
+//   <2> after merge operation, if any indexPath's tableFilter is not nil, we 
+//       should add indexPath’s tableFilter into IndexMergePath.tableFilters
+CreateIndexMergeIntersectionPath(partialPaths, tableFilters) {
+}
+
+```
+	
+### 3.3 Executor
+Graph bellow illustrates execution of IndexMerge scan. 
+
+![Execution model](./execution_model.png =400x200)
+
+Every index plan in `PhysicalIndexMergeLookUpReader` will start an `IndexWorker` to execute the IndexScan plan and send handles to AndOrWorker. AndOrWorker is responsible for doing set operations (and, or) to get final handles. Then `AndOrWoker` sends final handles to `TableWokers` to get rows from TiKV.
+
+We can take some tricks to make execution at pipeline mode without considering the order.
+
+
+(1)	IndexMergeIntersection
+
+
+ ![Execution model](./and.jpg =400x150)  

```

Comment:
img is not shown properly. and put them into subdirectory `imgs\`.

Reply:
ok, i will fix it.